Tree accessors and random generator

imane-aanna opened the pull request:
This branch aims to add accessors to the tree, and create a random generator of the tree.

imane-aanna committed:
Merge remote-tracking branch 'refs/remotes/origin/imane-aanna-tree-graph'

imane-aanna committed:
Merge branch 'master' of https://github.com/openalea-incubator/csMTG

imane-aanna committed:
Fixes #10

- Added a function NbVertices.
- Added a function NbChildre.
- Added getters for both Parent and children.
- Added tests for each of the new functions.

pradal commented on code/Tree.cs:
```diff
@@ -37,15 +37,90 @@ public class Tree
+            Dictionary<int, List<int>> temporary = children;
+
+            if (!temporary.ContainsKey(verticeId))
+                nbChildren = -1;
```
NbChildren has to be >= 0
In the case of the vertex do not belong to the Tree, return 0 or an error.
I would prefer 0.

imane-aanna committed:
Fixes #11

Added a function which enables to generate a tree when given the number
of vertices and maximum number of children per vertex

imane-aanna committed:
Fixes #14

- Removed the useless temporary attributes in some functions.
- NbChildren now returns 0 if the vertex doesn't exist. (Instead of -1)
- Parent(vid) now returns null in case the parameter doesn't exist.
(Instead of -999)
- Modified the tests accordingly

imane-aanna committed:
Fixes #16

Added three tests for the random tree generator :
- Verify that the number of vertices is the one specified in the
parameters (Used our function NbVertices).
- Verify that all vertices have less or equal than the function's
parameter. (Used our function : NbChildren())
- Verify that there is a coherence between the new attributes children
and parent. (Used our function: Children(vid) and Parent(vid)).

imane-aanna committed:
Close #15

Function RemoveVertex(vid, reparent_child):
- If the vid has children, reparent_child should be set to true in order
for the vertex to be removed. If it's set to false, the vertex won't be
removed.
- If the vid doesn't have children, it will be removed in all cases
unless it is the root.

imane-aanna commented:
@pradal : 
New changes have been added to this pull request:
- The returned values in case of errors are now changed.
- Unit tests have been created to validate the random tree generator.
- The function that enables to remove a vertex has been added.

pradal commented on code/Tree.cs:
```diff
@@ -37,15 +38,90 @@ public class Tree
+        /// <returns> Returns the total number of vertices. </returns>
+        public int NbVertices()
+        {
+            int nbVertices = parent.Count();
```
@imane-aanna : Use the Count method instead

imane-aanna replied:
I wrote in my side notes that it was a duplicate of the function Count() and I wanted to ask you about it indeed.

imane-aanna replied:
@pradal : I wrote in my side notes that it was a duplicate of the function Count() and I wanted to ask you about it indeed.

pradal commented on code/Tree.cs:
```diff
@@ -54,7 +130,7 @@ private int NewId()
                 id++;
             }
-            while (children.ContainsKey(id));
+            while (parent.ContainsKey(id));
```
@imane-aanna 
The complexity is O(n).
You can also use something like:
 max(parent)+1

pradal commented on code/Tree.cs:
```diff
@@ -125,10 +202,101 @@ void ReplaceParent(int parentId, int childId)
+        /// </summary>
+        /// <param name="nbVertices"> The number of tree's vertices </param>
+        /// <param name="nbChildren"> The maximum number for a vertex </param>
+        /// <returns> Returns the generated tree </returns>
```
@imane-aanna : I will add this algorithm outside this class in another module
like algorithm.cs

imane-aanna committed:
Fixes openalea-incubator/csMTG#17 openalea-incubator/csMTG#18 openalea-incubator/csMTG#19

- Deleted function NbVertices() and kept only Count().
- Change the way a new id is attributed.
- Remove the algorithm for generating a random tree into a new class
(Algorithm).
- Create separate tests for the new class Algorithm.

imane-aanna commented:
@pradal : Your remarks have been taken into account. In the last commit, I : 
- Deleted NbVertices() and kept Count().
- Changed the way a new id is attributed.
- Placed the random tree generator in a new Algorithm's class.

pradal commented:
Do not remove NbVertices function.
Just add
int NbVertices()
{ return Count(); }

imane-aanna committed:
Fixes openalea-incubator/csMTG#19

Rewrite the NbVertices function to directly use Count()

imane-aanna committed:
Fixes openalea-incubator/csMTG#20

- Created Unit Tests for the function RemoveVertices.
- Threw exceptions instead of Console messages for the unacceptable
cases.

imane-aanna commented:
@pradal  : Could you please check if everything is fine so that the branch is merged with the master and a new one is created for the property tree ?

pradal commented:
Good. I merge it
